packages/doenetml-worker-rust/lib-doenetml-core/src/utils/test_utils.rs

```diff
@@ -0,0 +1,57 @@
+use crate::{
+    components::prelude::{Dependency, StateVarMutableView, StateVarView, StateVarViewEnum},
+    dependency::DependencySource,
+    state::essential_state::EssentialDataOrigin,
+};
+
+/// Create a mutual view of a state variable along with a read only view
+pub fn create_view_pair<T: Default + Clone>(
+    value: T,
+    came_from_default: bool,
+) -> (StateVarMutableView<T>, StateVarView<T>) {
+    let mutable_view = StateVarMutableView::new_with_value(value, came_from_default);
+    let view = mutable_view.create_new_read_only_view();
+    (mutable_view, view)
+}
+
+pub fn create_essential_dependency<T: Default + Clone>() -> (Dependency, StateVarMutableView<T>)
+where
+    // make sure T is one of the types that we can convert from StateVarView<T> into StateVarViewEnum
+    StateVarViewEnum: From<StateVarView<T>>,
+{
+    let (essential, essential_view) = create_view_pair(T::default(), true);
+
+    let dependency_source = DependencySource::Essential {
+        component_idx: 0,
+        origin: EssentialDataOrigin::StateVar(0),
+    };
+
+    let dependency = Dependency {
+        source: dependency_source,
+        value: essential_view.into(),
+    };
+
+    (dependency, essential)
```

Comment:
The word `essential` still drives me nuts. This var isn't "an essential". It's a state var, so let's call it `state_var`.

Reply:
I like the derived versus essential/root/primitive distinction when dealing with the DAG of dependencies in core, as it is fundamental to how the DAG updates work. However, since we are trying to make the component development isolated from these complications, I see how it doesn't make sense to introduce such concepts in this domain.

There are two reasons why one would get "essential" variables in a state variable updater.
1. You specifically want to make the state variable depend on a local variable that doesn't depend on anything. Maybe for this case, you ask for a `DataQuery::LocalVariable` or something like that. (Currently, you are allowed only one of these per state variable and it will always have the type of your state variable, so not sure if there is a better name to communicate that.)
2. You ask for `DataQuery::Child` or `DataQuery::AttributeChild` and you get no results. So, you need somewhere to store your value if it were changed via `invert`. I had previously make these data queries return an single "essential" variable behind the scenes when there were no children in order to hide this detail from component authors. I recently removed that feature as it ended up getting confusing when to you had multiple data queries so that you had to ignore one of these automatically added variables. I'll think how I can bring that feature back while addressing that confusion, as it'd be nice to have this case work automatically without having component authors to code special cases.

Both of these cases were further complicated by the possibility of an `ExtendSource` that could create another dependency and require component developers to make a special case with handling these "essential" variable. I'll see if I can come up with a way to handle that automatically inside these other data queries so component authors don't have to deal with it, such as how you were suggesting earlier. (It will be a challenge to know what the automatic answer should be if, for example, there is both a `DataQuery::Child` and a `DataQuery::AttributeChild` in a single state variable.)

Reply:
I think I understand about 50%. Would `DataQuery::LocalVariable` be the same as `DataQuery::Literal(foo)`, where that would always just return the value `foo`?

Reply:
Not exactly, it's more like a "precursor" variable that a state variable developer can modify before calculating the actual value of a state variable. And then, the state variable developer would need to invert those calculations in `invert`.

In simple cases, such as for `sync_immediate_value` for `text_input`, no processing is done and the value is just passed through. (I need to create a general state variable case for that given that there is no logic to program there.)

But, in more complicated cases, such as `immediate_value`, that local value is just one piece of information needed to calculate the value. For example, the default from the local value is overridden by `prefill`.

I'm not really happy about the name. I'm not sure if `DataQuery::LocalValue` is better. It's certainly not static as it will typically be changed in `invert`. Although there's no way to enforce any behavior, the idea is that it is closely related to the value of `foo`.

Reply:
Decided to call it `DataQuery::PreliminaryValue` for now.